### DISCUSSION TITLE
CDI issues with Gradle multi-subproject in Kotlin

### DISCUSSION THREAD

**Discussion by TheBroach** [quarkusio/quarkus#43280] [Q&A] [answered]
Hi!
I am working on a simple blueprint-project for starting new Quarkus-Kotlin projects. I am following the Ports & Adapters design pattern and also want to follow the recommended approach from Gradle with multiple subprojects and convention plugins for sharing configurations across them. I also do not want to add Quarkus to a subproject if I do not need it, for example in the Domain subproject which should only contain pure business logic with little to no side effects.

However when applying the Quarkus Gradle plugin to several subprojects, I get an `UnsatisfiedResolutionException`. I have [created an MCVE](https://github.com/TheBroach/quarkus-gradle-cdi-mcve) for you to check out which should replicate my scenario. **_Note that repository definitions are omitted, see the README._**

If you see any potential improvements that could be made for my blueprint project, please point them out! :)

Thanks!
/John

[a comment by an automated account is not shown]

**Comment by mschorsch** [comment]
Quarkus 3.11.3 is rather old.

Could you try using Quarkus 3.14.2, Kotlin 2.0.10 and Gradle 8.10.1?

Hint: Do not use Quarkus 3.14.3 because it has several issues with Gradle which will be fixes with Quarkus 3.14.4.

**Reply by TheBroach** [reply]
Thanks for the tip! I bumped to those versions, unfortunately the issue remains. (The example repo is updated to those versions now.)

**Comment by mschorsch** [comment]
You have to add a beans.xml in meta-inf in each module (see https://quarkus.io/guides/cdi-reference#bean_discovery).

**Reply by TheBroach** [reply]
Yes, but I was under the impression that it was enough to add the Jandex Gradle plugin? (https://quarkus.io/guides/cdi-reference#how-to-generate-a-jandex-index)
Do I need to configure it in a special way since I am not using the Quarkus plugin in all subprojects? I put the Jandex plugin in my kotlin-conventions plugin since it needs to be present even in the subprojects where the Quarkus plugin is not present (if I understood it correctly).

**Reply by TheBroach** [reply]
Any thought on Jandex plugin configuration from @geoand ?

**Reply by geoand** [reply]
That's one for @Ladicek

**Reply by Ladicek** [reply]
I know next to nothing about the Jandex Gradle plugin(s). If the resulting JAR contains a file `META-INF/jandex.idx`, then the Jandex plugin is configured correctly (and indeed `beans.xml` shouldn't be necessary in that case). If it does not, well, there's the issue.

**Reply by aloubyansky** [reply]
https://github.com/quarkusio/quarkus/blob/main/devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java#L669-L677 is the only code we have to set dependencies on Jandex.
@TheBroach it fails `quarkusDev` or a build?

**Reply by Ladicek** [reply]
The reproducer seems to configure Jandex on a few places (e.g. https://github.com/TheBroach/quarkus-gradle-cdi-mcve/blob/main/gradle/build-logic/kotlin-conventions/src/main/kotlin/kotlin-conventions.gradle.kts#L10), but I'm not sure if that's enough for all the projects to actually apply the Jandex plugin :shrug:

**Reply by TheBroach** [reply]
Thanks guys for taking a look!

@aloubyansky It fails `build`, but when running `quarkusDev` I noticed that the logs only declare `INFO  [io.quarkus] (Quarkus Main Thread) driven-adapters unspecified on JVM (powered by Quarkus 3.14.2) started in 0.771s. `, the same log entry for the `driving-adapters` subproject is missing (where the Quarkus plugin is also applied via the quarkus-conventions plugin). That is also the subproject for which the build fails. Maybe that's saying something?

@Ladicek Thanks for the suggestion of inspecting the JAR, will do that. (Still a junior dev, so not too familiar yet with compiling/JARs/CDI/the-whole-shebang...)

**Reply by TheBroach** [reply]
I checked the JARs `domain.jar`, `driven-adapters.jar` and `driving-adapters.jar`, they all had the `META-INF/jandex.idx` file...

**Reply by Ladicek** [reply]
Then I guess you're using the Quarkus plugin incorrectly. It makes no sense to use the Quarkus plugin on arbitrary modules -- it should only be present in the module that assembles the final application. (This module should have a dependency on all the other modules that constitute the application.)

**Reply by TheBroach** [reply]
Hm... Couldn't there be Quarkus features that I would want to use in a module designated for other things than to assemble the whole application, such as an "infrastructure"/"driven adapter" module where I would like to use some convenient REST client or GRPC client features from Quarkus?

**Reply by Ladicek** [reply, marked as answer]
That is correct, and you should feel free to use Quarkus _dependencies_ in those modules. But you don't need to use the Quarkus _Gradle plugin_ there.

**Reply by TheBroach** [reply]
Oooh, ok... Thanks, got it working now using a separate "application" module. I will update my MCVE when I've cleaned it up a bit!

Still curious about things that are sort of related to Quarkus-Gradle-tasks, such as generating code for a GRPC client in that potential "infrastructure"/"driven adapter" module... Will continue toying around with that, can post a new question if that proves too big of a challenge as well!
